bpo-46730: Add more info to @property AttributeError messages

Alex-Blade opened the pull request:
The property decorator raises AttributeErrors with messages
that neither help to identify the problem nor lead to a possible solution.
This PR proposes changes in error messages to address the issue.

<!-- issue-number: [bpo-46730](https://bugs.python.org/issue46730) -->
https://bugs.python.org/issue46730
<!-- /issue-number -->

Commit:
bpo-46730: Make @property AttributeError unique and sensible

The property decorator raises AttributeErrors with messages
that neither help to identify the problem nor lead to a possible solution.
This commit proposes changes in error messages to address the issue.

Commit:
bpo-46730: Add more details on property origin to the AttributeError

This is an improvement to the initial patch for the bpo-46730.
It adds additional information on class that owns the property
that caused an AttributeError.

Commit:
bpo-46730: Make AttributeError show self.__qualname__ instead of owner type

Commit:
bpo-46730: Adjusted documentation according to the latest changes

sweeneyde commented on Lib/test/test_property.py:
```diff
@@ -314,35 +314,35 @@ class _PropertyUnreachableAttribute:
 
-    def _format_exc_msg(self, msg):
-        return self.msg_format.format(msg)
+    def _format_exc_msg(self, *msg):
```
This change looks no longer necessary?

Alex-Blade replied:
Yep

sweeneyde commented on Lib/test/test_property.py:
```diff
@@ -314,35 +314,35 @@ class _PropertyUnreachableAttribute:
 
 class PropertyUnreachableAttributeWithName(_PropertyUnreachableAttribute, unittest.TestCase):
-    msg_format = "^{} 'foo'$"
+    msg_format = "^property 'foo' in object of type 'PropertyUnreachableAttributeWithName.cls' {}$"
```
Since this is a regex, I think it should technically have that period escaped like `r"...Name\.cls"` or `"...Name\\.cls"`

Alex-Blade replied:
Indeed, fixed

sweeneyde commented on Objects/descrobject.c:
```diff
@@ -1586,9 +1586,14 @@ property_descr_get(PyObject *self, PyObject *obj, PyObject *type)
         if (gs->prop_name != NULL) {
-            PyErr_Format(PyExc_AttributeError, "unreadable attribute %R", gs->prop_name);
+            PyErr_Format(PyExc_AttributeError,
+                         "property %R in object of type %R has no getter",
```
I'm not sure "in" is the right preposition here (and likewise in the other spots).
Maybe "for" or "on" or "of" instead?

Alex-Blade replied:
Yeah, "of" is actually the best in my opinion but I didn't like the tautology

UPD. Though now as I think of it, seems like it'd be fine with "of"

sweeneyde commented on Lib/test/test_property.py:
```diff
@@ -314,35 +314,35 @@ class _PropertyUnreachableAttribute:
 
 class PropertyUnreachableAttributeNoName(_PropertyUnreachableAttribute, unittest.TestCase):
-    msg_format = "^{}$"
+    msg_format = "^property in object of type 'PropertyUnreachableAttributeNoName.cls' {}$"
```
Likewise, escape the period.

Alex-Blade replied:
Indeed, fixed

sweeneyde reviewed (changes requested):
This is mostly looking good, just some minor nits remaining.

sweeneyde commented on Objects/descrobject.c:
```diff
@@ -1611,18 +1616,24 @@ property_descr_set(PyObject *self, PyObject *obj, PyObject *value)
+                        "property %R in object of type %R has no setter",
+                        gs->prop_name,
+                        PyType_GetQualName(Py_TYPE(obj)));
+        } else if (obj != NULL) {
```
Another minor thing: [PEP 7](https://www.python.org/dev/peps/pep-0007/) says to put `else` on the next line.

Alex-Blade replied:
Fixed

Commit:
bpo-46730: Fix regex dot in tests

Properly escape dot character and return `_format_exc_msg` function to its initial state.

Commit:
bpo-46730: Make changes comply with PEP 7

Commit:
bpo-46730: Change preposition from "in" to "of" in error

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

sweeneyde reviewed (approved):
LGTM. I'll leave it open for a day or so to see if there are any more comments, but after that, I can merge.

If you want, you can add your name to MISC/ACKS as well.

Commit:
Update ACKS

eryksun commented on Objects/descrobject.c:
```diff
@@ -1586,9 +1586,15 @@ property_descr_get(PyObject *self, PyObject *obj, PyObject *type)
+            PyErr_Format(PyExc_AttributeError,
+                         "property %R of object of type %R has no getter",
+                         gs->prop_name,
+                         PyType_GetQualName(Py_TYPE(obj)));
```
I think "property %R of object of type %R" is a bit awkward. I'd prefer "property %R of %R objects has no getter".

sweeneyde replied:
Yeah, there are lots of different colors this bikeshed could be painted. I think the singular "object" is more consistent with other error messages than the plural "objects", but other than that, I like your suggestion. @Alex-Blade what do you think?

Alex-Blade replied:
To be honest, I don't understand what does it mean from the first glance. Like is it the same as "Property %R that appears in all objects of type %R..."?

Alex-Blade replied:
Tested out it a bit - strongly disagree with plural form, but in general I like it, so pushed "property %R of %R object has no getter"

Commit:
bpo-46730: Remove "of type of" from the message

Commit:
bpo-46730: Change the order of words

sweeneyde commented on Doc/howto/descriptor.rst:
```diff
@@ -1456,7 +1456,7 @@ attributes stored in ``__slots__``:
     Traceback (most recent call last):
         ...
-    AttributeError: can't set attribute
+    AttributeError: no setter was defined for property
```
Should this be updated?

Alex-Blade replied:
Yeah indeed, but I'm not sure how to run these tests. Neither patchcheck nor full test fails, meaning it's some other command

sweeneyde replied:
I ran by hand in a separate file to double-check:

```python
class Immutable:
    """
    >>> mark = Immutable('Botany', 'Mark Watney')
    >>> mark.dept
    'Botany'
    >>> mark.dept = 'Space Pirate'
    Traceback (most recent call last):
        ...
    AttributeError: property 'dept' of 'Immutable' object has no setter
    >>> mark.location = 'Mars'
    Traceback (most recent call last):
        ...
    AttributeError: 'Immutable' object has no attribute 'location'
    """
    __slots__ = ('_dept', '_name')          # Replace the instance dictionary

    def __init__(self, dept, name):
        self._dept = dept                   # Store to private attribute
        self._name = name                   # Store to private attribute

    @property                               # Read-only descriptor
    def dept(self):
        return self._dept

    @property
    def name(self):                         # Read-only descriptor
        return self._name

import doctest
doctest.testmod()
```

Commit:
bpo-46730: Fix doctest for property
